### PR TITLE
server: make leader can lost lease as soon as possible

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

Fix https://github.com/tikv/pd/issues/3697
When the store's number is 180. we resign pd leader  may need one minute or more to finish elect the new leader:
image

details replay: https://gist.github.com/disksing/a1bf78007b674da7e15b52a9e25eb3a5 

### What is changed and how it works?

make leader can be lost lease as soon as possible

### Check List

Tests

- Integration test

### Release note

```release-note
Fix the issue that leader re-election be slow when there are many stores
```

### CODE REVIEW

File: tests/server/member/member_test.go [hunk 1]
```diff
@@ -238,12 +239,31 @@ func (s *memberTestSuite) TestLeaderResign(c *C) {
...
+	c.Log("leader2:", leader2)
+	c.Assert(leader2, Not(Equals), leader1)
+}
+
 func (s *memberTestSuite) waitLeaderChange(c *C, cluster *tests.TestCluster, old string) string {
 	var leader string
 	testutil.WaitUntil(c, func(c *C) bool {
 		leader = cluster.GetLeader()
 		if leader == old || leader == "" {
-			time.Sleep(time.Second)
```
Comment (lhy1024) [inline tests/server/member/member_test.go:246]: What will be the side effects of removing it?

Comment (nolouch) [reply tests/server/member/member_test.go:246]: Actually `testutil.WaitUntil` will sleep in the retry with `100 * time. Millisecond` interval, It's already enough.

File: server/server.go [hunk 2]
```diff
@@ -1213,6 +1214,7 @@ func (s *Server) campaignLeader() {
 	ctx, cancel := context.WithCancel(s.serverLoopCtx)
 	defer cancel()
 	defer s.member.ResetLeader()
```
Comment (JmPotato) [inline server/server.go:1216]: These two lines should be removed now.

File: server/server.go [hunk 3]
```diff
@@ -1256,6 +1258,11 @@ func (s *Server) campaignLeader() {
 		return
 	}
 	s.member.EnableLeader()
+	defer func() {
+		s.member.ResetLeader()
+		// as soon as cancel the keepalive.
+		cancel()
+	}()
```
Comment (JmPotato) [inline server/server.go:1265]: I think the TSO resetting better should be moved to here too.

File: server/server.go (hunk 2 above)
Comment (rleungx) [reply server/server.go:1216]: I think we still need to cancel the context when there is a problem before we reach the below defer function.

Comment (nolouch) [reply server/server.go:1216]: yes, some branches will return early.

File: server/server.go (hunk 3 above)
Comment (nolouch) [reply server/server.go]: Once leadership is reset and tso will stop service.

**Comment by nolouch** [comment]
PTAL @JmPotato  @rleungx

File: server/server.go (hunk 2 above)
Comment (rleungx) [reply server/server.go:1216]: But it will be reset and canceled twice.

File: tests/server/member/member_test.go [hunk 4]
```diff
@@ -238,12 +239,31 @@ func (s *memberTestSuite) TestLeaderResign(c *C) {
...
+
+	err = cluster.RunInitialServers()
+	c.Assert(err, IsNil)
+
+	leader1 := cluster.WaitLeader()
+	addr1 := cluster.GetServer(leader1).GetConfig().ClientUrls
+
+	err = failpoint.Enable("github.com/tikv/pd/server/raftclusterIsBusy", `pause`)
+	defer failpoint.Disable("github.com/tikv/pd/server/raftclusterIsBusy")
+	c.Assert(err, IsNil)
```
Comment (JmPotato) [inline tests/server/member/member_test.go:255]: `defer` should come after the `err` check.

File: server/server.go [hunk 5]
```diff
@@ -1255,7 +1255,16 @@ func (s *Server) campaignLeader() {
 		log.Error("failed to sync id from etcd", errs.ZapError(err))
 		return
 	}
+	// EnableLeader to accept the remaining service, such as GetStore, GetRegion.
 	s.member.EnableLeader()
+	// Check the cluster dc-location after the PD leader is elected
```
Comment (JmPotato) [inline server/server.go:1260]:
```suggestion
	// Check the cluster dc-location after the PD leader is elected.
```

**Review by JmPotato — APPROVED** [review]
Rest LGTM.

File: server/server.go (hunk 2 above)
Comment (nolouch) [reply server/server.go:1216]: yes, it's valid.

**Review by rleungx — APPROVED** [review]

**Pull request merged**